core/src/main/java/org/renjin/primitives/S3.java

```diff
@@ -258,6 +268,314 @@ public static SEXP tryDispatchFromPrimitive(Context context, Environment rho, Fu
+    // signature length is. This might be longer the length of arguments and #ANY should be used for missing
+    // arguments. In case signature is shorted than the number of arguments we don't need to evaluate the extra
+    // arguments.
+  
+    int signatureLength = computeSignatureLength(genericMethodEnvironment, groupMethodEnvironment);
+  
+    ArrayList<MethodRanking> possibleSignatures = generatePossibleSignatures(context, rho, args, signatureLength);
+  
+    Collections.sort(possibleSignatures);
+  
+    ArrayList<SelectedMethod> selectedMethods = findMatchingMethods(context, genericMethodEnvironment, possibleSignatures, group, opName);
+  
+    if(selectedMethods.isEmpty()) {
+      throw new EvalException("object of type 'S4' is not subsettable");
+    }
+    SelectedMethod method = selectedMethods.get(0);
+
+//     if selected method is from Group or if its from standard generic but distance is > 0
+//     returned function environment is populated with the following metadata objects:
+//     - .defined:  signature method
+//     - .target:   signature target
+//     - .Generic:
+//     - .Method:   method definition
+//     - .Methods:  function call
+//     - e1:        arg1
+//     - e2:        arg2
+//
+//     otherwise only e1 and e2.
+  
+    if (method.getGroup().equals("generic")) { /*  && method.getDistance() == 0 */
+      return context.evaluate(new FunctionCall(selectedMethods.get(0).getFunction(), args));
+    } else {
+      SEXP objDefined = generateDefinedObjectMetadata(context, method);
+      SEXP generic = generateObjectGenericMetadata(opName);
+    
+      Map<Symbol, SEXP> metadata = new HashMap<>();
+      metadata.put(Symbol.get(".defined"), objDefined);
+      metadata.put(Symbol.get(".Generic"), generic);
+      metadata.put(Symbol.get(".Method"), method.getFunction());
+      metadata.put(Symbol.get(".Methods"), Symbol.get(".Primitive(\"" + opName +"\")"));
+      metadata.put(Symbol.get(".target"), objDefined);
+    
+      return ClosureDispatcher.apply(context,rho, new FunctionCall(method.getFunction(), args), (Closure) method.getFunction(), args, metadata);
+    }
+  }
+  
+  private static SEXP generateObjectGenericMetadata(String opName) {
+    SEXP generic = Symbol.get(opName);
+    generic.setAttribute("package", StringVector.valueOf("base"));
+    return generic;
+  }
+  
+  private static SEXP generateDefinedObjectMetadata(@Current Context context, SelectedMethod method) {
+    StringVector.Builder objDefinedSig = new StringVector.Builder();
+    String[] objDefinedClasses = method.getInputSignature().split("#");
+    for(int i = 0; i < objDefinedClasses.length; i++) {
+      objDefinedSig.add(objDefinedClasses[i]);
+    }
+    SEXP objDefined = Symbol.get(deparse(context, objDefinedSig.build(), 256, false, 0, 1));
+    objDefined.setAttribute(".Names", ((Closure) method.getFunction()).getFormals().getNames());
```

Comment:
Are you sure this attribute is called ".Names"?? Note that `deparse(c(a=1))` evaluates to `"structure(1, .Names = \"a\")"` but doesn't mean the attribute is called ".Names", it's just a quirk of the structure function:

```
> deparse(c(a=1))
[1] "structure(1, .Names = \"a\")"
> attributes(c(a=1))
$names
[1] "a"
```